mailpile/plugins/tags.py

```diff
@@ -361,29 +369,45 @@ def _do_tagging(self, ops, msg_ids, conversations, save=True, auto=False):
                 if conversation:
                     rv['conversations'] = True
 
+                ignored = False
                 tag_id = tag._key
+                tag_cfg = tag
                 tag = tag.copy()
                 tag["tid"] = tag_id
                 if op[0] == '-':
-                    removed = idx.remove_tag(self.session, tag_id,
-                                             msg_idxs=msg_ids,
-                                             conversation=conversation)
-                    rv['untagged'].append((tag, sorted([b36(i)
-                                                        for i in removed])))
+                    if force or tag_cfg['flag_allow_del']:
+                        removed = idx.remove_tag(self.session, tag_id,
+                                                 msg_idxs=msg_ids,
+                                                 conversation=conversation)
+                        rv['untagged'].append(
+                            (tag, sorted([b36(i) for i in removed])))
+                    else:
+                        rv['ignored'].append((op, tag))
+                        ignored = True
                 else:
-                    added = idx.add_tag(self.session, tag_id,
-                                        msg_idxs=msg_ids,
-                                        conversation=conversation)
-                    rv['tagged'].append((tag, sorted([b36(i)
-                                                      for i in added])))
+                    if force or tag_cfg['flag_allow_add']:
+                        added = idx.add_tag(self.session, tag_id,
+                                            msg_idxs=msg_ids,
+                                            conversation=conversation)
+                        rv['tagged'].append(
+                            (tag, sorted([b36(i) for i in added])))
+                    else:
+                        rv['ignored'].append((op, tag))
+                        ignored = True
+
                 # Record behavior
-                if len(msg_ids) < 15:
+                if len(msg_ids) < 15 and not ignored:
                     for t in self.session.config.get_tags(type='tagged',
                                                           default=[]):
                         idx.add_tag(self.session, t._key, msg_idxs=msg_ids)
             else:
                 self.session.ui.warning('Unknown tag: %s' % op)
 
+
+        if rv['ignored'] and (len(rv['tagged']) == len(rv['untagged']) == 0):
+            self.event.private_data['ignored'] = rv['ignored']
+            return self._error(_('Nothing Happened'), result=rv)
+
```

Comment:
I struggled with that a bit; I agree this would be a useful enhancement for the future.

In practice this is a pretty rare event and the UI behaviour will actually be relatively sane: This only comes up (AFAICT) if the user is viewing a compound result (so a union of Tag A and Tag B), where Tag A is removable but Tag B is not. If the user then clicks the "untag" button or drags the message somewhere else, the result at the backend is that the removable tag (Tag A) will be stripped off the messages (but not Tag B), and the result in the user interface is the messages will disappear from view because they're no longer in the union.

As you point out, the user won't be informed that Tag B wasn't stripped, but everything else will behave correctly. In general, explaining to the user in the UI that some tags are removable but others are not is a hard problem and I could use some help from someone with visual design skills there.